### ISSUE TITLE
Exported SVG way too big

### ISSUE THREAD

**Issue by Gianarci** [gerbv/gerbv#236]
I wanted to report that the tool exports SVGs that are poorly optimized and often too heavy to render for the browser. In one board, it exported a 30Mb file. By optimizing it with a tool (SVGO) the file was reduced to 1.8Mb, with the same graphic rendering. The main problem is that approximately 112000 paths are generated (with absolute coordinates) and identical styles, while SVGO reduces them to 206 (merging them), transforming them into relative coordinates, applying any transforms, and this greatly optimizes the rendering phase. However, given its size, the aforementioned tool takes 9 minutes to perform the optimization and is therefore an unacceptable solution. Is there a way to mitigate this problem?

**Comment by spe-ciellt** [comment]
I have not looked into this problem deeply, but my feeling is that translation from Gerber to SVG is done on an element basis. So one Gerber element is converted to one SVG element. If the original Gerber file have many small elements the generated SVG file will also have many small elements. SVG is XML based and therefore the file will get big.

**Comment by Gianarci** [comment]
Yes, the conversion to SVG is quite naive and can be significantly optimized. Often (depending on the version of the CairoSVG library!), the output is a long list of SVG primitives with a long, replicated inline style:

```
<path style="fill-rule:evenodd;fill:rgb(72.156863%, 45.098039%, 20%);fill-opacity:1" d="M 72.730469 0.511719 L 88.078125 0.511719 L 88.078125 74.460938 L 16.03125 74.460938 ... "/>
<path style="fill-rule:evenodd;fill:rgb(72.156863%, 45.098039%, 20%);fill-opacity:1" d="..."/>
<path style="fill-rule:evenodd;fill:rgb(72.156863%, 45.098039%, 20%);fill-opacity:1" d="..."/>
<path style="fill-rule:evenodd;fill:rgb(72.156863%, 45.098039%, 20%);fill-opacity:1" d="..."/>
<path style="fill-rule:evenodd;fill:rgb(72.156863%, 45.098039%, 20%);fill-opacity:1" d="..."/>
<path style="fill-rule:evenodd;fill:rgb(72.156863%, 45.098039%, 20%);fill-opacity:1" d="..."/>
<path style="fill-rule:evenodd;fill:rgb(72.156863%, 45.098039%, 20%);fill-opacity:1" d="..."/>
<path style="fill-rule:evenodd;fill:rgb(72.156863%, 45.098039%, 20%);fill-opacity:1" d="..."/>
<path style="fill-rule:evenodd;fill:rgb(72.156863%, 45.098039%, 20%);fill-opacity:1" d="..."/>
<path style="fill-rule:evenodd;fill:rgb(72.156863%, 45.098039%, 20%);fill-opacity:1" d="..."/>
<path style="fill-rule:evenodd;fill:rgb(72.156863%, 45.098039%, 20%);fill-opacity:1" d="..."/>
<path style="fill-rule:evenodd;fill:rgb(72.156863%, 45.098039%, 20%);fill-opacity:1" d="..."/>
```

This can be optimized by creating a class instead and replacing the style with it:

<path classstyle="fill-rule:evenodd;fill:rgb(72.156863%, 45.098039%, 20%);fill-opacity:1" d="M 72.730469 0.511719 L 88.078125 0.511719 L 88.078125 74.460938 L 16.03125 74.460938 ... "/>
<path style="fill-rule:evenodd;fill:rgb(72.156863%, 45.098039%, 20%);fill-opacity:1" d="..."/>

In addition, you can also shorten the paths by merging some of them that are continuous and the number of decimal digits can be reduced (from 6 to 3 for example).

These optimizations can reduce the size of the SVG up to a factor of 10, and therefore also the rendering time.

Unfortunately, in newer versions of CairoSVG the inline style has been replaced with the corresponding SVG attributes, making it more difficult to apply the first optimization:

```
<path fill-rule="evenodd" fill="rgb(72.156863%, 45.098039%, 20%)" fill-opacity="1" d="M 72.730469 0.511719 L 88.078125 0.511719 L 88.078125 ... "/>
<path fill-rule="evenodd" fill="rgb(72.156863%, 45.098039%, 20%)" fill-opacity="1" d="..."/>
<path fill-rule="evenodd" fill="rgb(72.156863%, 45.098039%, 20%)" fill-opacity="1" d="..."/>
<path fill-rule="evenodd" fill="rgb(72.156863%, 45.098039%, 20%)" fill-opacity="1" d="..."/>
<path fill-rule="evenodd" fill="rgb(72.156863%, 45.098039%, 20%)" fill-opacity="1" d="..."/>
<path fill-rule="evenodd" fill="rgb(72.156863%, 45.098039%, 20%)" fill-opacity="1" d="..."/>
<path fill-rule="evenodd" fill="rgb(72.156863%, 45.098039%, 20%)" fill-opacity="1" d="..."/>
<path fill-rule="evenodd" fill="rgb(72.156863%, 45.098039%, 20%)" fill-opacity="1" d="..."/>
<path fill-rule="evenodd" fill="rgb(72.156863%, 45.098039%, 20%)" fill-opacity="1" d="..."/>
<path fill-rule="evenodd" fill="rgb(72.156863%, 45.098039%, 20%)" fill-opacity="1" d="..."/>
<path fill-rule="evenodd" fill="rgb(72.156863%, 45.098039%, 20%)" fill-opacity="1" d="..."/>
<path fill-rule="evenodd" fill="rgb(72.156863%, 45.098039%, 20%)" fill-opacity="1" d="..."/>
```

However, this problem can be solved by replacing the attributes with inline styling, although it would be nice if the library allowed you to choose between the 2 representations...

**Comment by eyal0** [comment]
The shapes in gerbv need to be kept separate so that they can be manipulated independently, like if you want to delete or move just one of them or something.

Merging them is a complicated process that will necessarily cause a loss in precision due to floating point.

I expect that the number of paths in the SVG should be the same as the number of shaped in the input gerber file.  Is that right?

**Comment by Gianarci** [comment]
It depends entirely on the intended use. If I just need to display a vectorial image, I don't care about keeping them separate. Instead, rendering times (and memory requirements) become significant.

**Comment by eyal0** [comment]
Right.  So where would that merging occur?  Either gerbv gets a library to merge shapes or you continue to use the SVGO.  `pcb2gcode` has a shape merger in it, too, that uses gerbv.  I tried to make it fast.  I don't know if it is faster than SVGO but I could measure that for you if you are interested.

I think that merging gerber shapes into a gerber file is probably too difficult.  It would be more reasonable to convert gerber to svg and then merge them, as you are doing with SVGO, or merge gerber shapes into SVG as they are read, which is what pcb2gcode is doing.  I should think that the latter is faster because it doesn't need to do an intermediate step.

Like I said, if you have a file for me, I can instrument pcb2gcode to measure the time to merge.  And if it's good, maybe it's worthwhile to make this into a standalone library?

**Comment by eyal0** [comment]
This is the entrypoint for the code: https://github.com/pcb2gcode/pcb2gcode/blob/master/gerberimporter.cpp#L752

It converts a gerber file into:

* many polygons that don't overlap and may have holes in them
* many paths, each with a width.  Optionally, you can choose to have all those lines converted into shapes and merged into the polygons.  (For some uses, it can be useful to have the lines separate from the shapes.)

**Comment by rampageservices** [comment]
I've submitted PR #317 with an implementation that addresses this issue by adding a custom SVG writer that bypasses Cairo's SVG surface entirely.

**Approach:** Instead of going through Cairo (which emits one `<path>` per Gerber primitive with repeated inline styles and 6-digit precision), the new writer walks gerbv's internal data structures directly and emits optimized SVG:

- **CSS class deduplication** — unique fill/stroke combinations are collected into a `<style>` block and referenced via `class="c0"`, `class="c1"`, etc. A typical board has <20 unique styles vs 112,000 inline style repetitions.
- **SVG-native elements** — `<circle>`, `<rect>`, `<line>` instead of generic `<path>` where applicable.
- **Reduced precision** — 3 decimal places instead of 6 (0.00035mm resolution at 72 DPI, well below visual threshold), with trailing zeros stripped.
- **`<g>` grouping** — nested groups carry shared attributes and transforms.

**Results on test files:**
- 42–69% file size reduction across test Gerbers
- Zero test regressions (94/94 tests pass identically)
- `--svg-cairo` CLI flag and GUI checkbox available to fall back to the legacy Cairo SVG output

The existing Cairo SVG path is fully preserved behind the `--svg-cairo` flag, so there's no risk of regression for users who prefer the old output.

I'd appreciate any testing and feedback — especially from anyone with large or complex board files who can compare the optimized SVG output against the Cairo version. Visual fidelity, file size, and compatibility with downstream tools (Inkscape, browsers, etc.) are all useful data points.
